LowLisp Features.md

```diff
@@ -1,3 +1,37 @@
 # Features Necessary for LowLisp
 
 This document lists everything that will be needed in LowLisp, in order to implement HighLisp.
+
+
+## Language
+
+### Functions
+
+* Are not first class, but experience only a small amount of discrimination. There are function pointers (as in C), and attempting to assign a function to a variable just stores its respective address.
+* Cannot be redefined. However, code can modify a function pointer that will be called in the future. This can be used to implement function redefinition in higher-level languagues.
+* Lack reflection and metaprogramming. Code cannot see which S-expressions make up a function, and cannot modify the function in any way other than overwriting its machine code in memory (probably a bad idea).
+* Must have their parameters and return values defined completely (the generic pointer counts as a defined type).
+
+### Variables
+
+* Are lexically scoped.
+* Are allocated on the stack when declared inside of a function.
+* Are allocated on the heap when declared outside of a function.
+* Must have a type that is known at compile-time, which must make include their size in memory.
+
+### Pointers
+
+* By default are reference-counted, and automatically deallocate their target once the last of their kind falls into oblivion.
+* Are also available in raw form.
+* Must point to a type known at compile time.
+* May point to a tagged union (type :go, for “generic object”).
+
+## Library
+
+### Memory allocation
+
+* Equivalents to C's malloc(), free(), and string functions.
```

Comment:
These should also be a) secure by default, and b) guarantee allocated data is zero'd.

Reply:
Alright. The latter makes sense. We should be more specific with “secure”.

Reply:
They should not leak data in case the OS does not zero RAM before giving it to another program.

Reply:
Ah, so free() should zero data before giving it back to the OS.

Reply:
Yes, additionally, we should make sure we always use the secure library functions provided by the OS, just in case. Though that isn't really a language spec issue.